src/pykeen/training/training_loop.py

```diff
@@ -78,12 +80,12 @@ def _get_optimizer_kwargs(optimizer: Optimizer) -> Mapping[str, Any]:
     return optimizer_kwargs
 
 
-class TrainingLoop(ABC):
+class TrainingLoop(Generic[SampleType, BatchType], ABC):
     """A training loop."""
 
-    training_instances: Optional[Instances]
+    training_instances: Optional[Instances[SampleType]]
```

Comment:
does this even need to be a variable stored in the training loop instance? It's both set and only used in `TrainingLoop._train()`. Could it just be a local variable in that function?

Reply:
Hmm, it's set in `TrainingLoop.train()`, so maybe this could be passed around as another variable?

Reply:
It might be able to replace the `triples_factory` passed to `train` 🤔

Reply:
Turns out it's not so easy to replace the triples_factory unless we want to modify the instances to explicitly store a bit more information, like `num_triples` and `mapped_triples` (these aren't available in LCWA instances)